### ISSUE TITLE
One way edge to same type cause the field become unique

### ISSUE THREAD

**Issue by wenerme** [ent/ent#1341]
- [x] The issue is present in the latest release.
- [x] I have searched the [issues](https://github.com/facebook/ent/issues) of this repository and believe that this is not a duplicate.

## Current Behavior 😯

```go
type AuditorMixin struct {
	mixin.Schema
}

func (AuditorMixin) Fields() []ent.Field {
	return []ent.Field{
		field.Int("createdByID").Optional().Nillable(),
	}
}
func (AuditorMixin) Indexes() []ent.Index {
	return []ent.Index{
		index.Fields("createdByID"),
	}
}
func (AuditorMixin) Edges() []ent.Edge {
	return []ent.Edge{
                // one way reference
		edge.To("createdBy", User.Type).Field("createdByID").Unique(),
	}
}

type User struct {
	ent.Schema
}

func (User) Mixin() []ent.Mixin {
	return []ent.Mixin{
		AuditorMixin{},
	}
}
```

The generated migrate

```go
	UsersColumns = []*schema.Column{
		{Name: "id", Type: field.TypeInt, Increment: true},
                // Should not Unique
		{Name: "created_by_id", Type: field.TypeInt, Unique: true, Nullable: true},
	}
```

> Apply mixin to different type is ok

## Expected Behavior 🤔

user `created_by_id` is not Unique

## Steps to Reproduce 🕹

https://github.com/wenerme/ent-bug

## Your Environment 🌎

| Tech        | Version |
| ----------- | ------- |
| Go          | 1.16  |
| Ent         | v0.7.0   |
| Database    | n/a   |
| Driver      | n/a      |

**Comment by a8m** [comment]
Hey @wenerme and thanks for raising up this issue.
What type of relationship are you trying to define? O2O, O2M?

**Comment by wenerme** [comment]
@a8m Every entity has a `createdBy` field, referenced to `User`, this is a one way relationship on mixin. Because the user side don't need to know this edge.

**Comment by a8m** [comment]
When defining an edge (relationship) from a type to itself, there are 3 options:

1\. [O2O Bidi](https://entgo.io/docs/schema-edges/#o2o-bidirectional) - a bidirectional edge from a type to itself. For example, a user has a spouse, and this spouse has a user. This column must be unique because otherwise, it will violate the relationship constraint. 
2\. [O2O Same Type](https://entgo.io/docs/schema-edges/#o2o-same-type) - O2O from a type to itself. See the linked-list in the example. A node has prev and next, and this must be unique because otherwise, it won't be O2O.
3\. [O2M Same Type](https://entgo.io/docs/schema-edges/#o2m-same-type) (your case btw). A user can create many users and a user can be created by 1 user. So, in your case, the edge will look as follows:
  ```go
  edge.To("created", User.Type).
	From("creator").
	Unique().
	Field("creator_id"),
  ```

**Comment by wenerme** [comment]
@a8m I updated the repo add another entity, seems not works.

The edge

```go
edge.To("createdBy", User.Type).From("creator").Field("createdByID").Unique(),
```
 generate error code

![image](https://user-images.githubusercontent.com/1777211/111437460-17166d80-873e-11eb-9987-f8c7fa2ed680.png)

The department failed to generated with 

```
running ent codegen: entc/gen: assoc-inverse edge allowed only as o2o relation of the same type
```

**Comment by wenerme** [comment]
Fixed the first problem, `edge.To("createdBy", User.Type).From("creator").Field("createdByID").Unique(),`

But does this mean this edge can not be defined on Mixin ?

**Comment by a8m** [comment]
Please use this configuation: 

```go
edge.To("created", User.Type).
  From("creator").
  Unique().                
  Field("creator_id"),
```

Or, in your case, this:

```go
edge.To("created", User.Type).
    From("created_by").
    Unique().
    Field("created_by_id"),
```

If it's hard to follow this way, you can break it to 2 different edges:
```go
edge.To("created", User.Type), // Many.
edge.From("created_by", User.Type). // One.
  Ref("created").
  Unique().
  Field("created_by_id"),
```

**Comment by a8m** [comment]
> But does this mean this edge can not be defined on Mixin ?

If it's defined in a Mixin, it means that it's an O2M edge for 2 different types.

**Comment by a8m** [comment]
OK, I get your problem - this mixed-edge works like an O2M edge when you embed it in different schemas, but when you embed it in the `User` schema, you get the O2O relationship - right? If that's the case, just don't use this Mixin in the `User` schema, and use it in all others.

**Comment by wenerme** [comment]
@a8m Yes, ok.

**Comment by a8m** [comment]
Cool. Closing this issue then, but please feel free to re-open it if you have troubles with it, or ping me on Slack 😃 

Thanks!

**Comment by wenerme** [comment]
@a8m By define the different edge on user using

```go
edge.To("created", User.Type).
  From("creator").
  Unique().                
  Field("creator_id"),
```

will create a `created` field on user, I guess there is no way to avoid this before allowing `edge.From("creator", User.Type).Unique().Field("creator_id")` ?
